doc/tsm/tsi.md

```diff
@@ -0,0 +1,87 @@
+
+
+## Block types
+
+The file contains multiple block types:
+
+- Series Block - This block stores a series block ID and a list of series ids that map to series keys. The list is sorted by series key. The series id is only used internally within the index file and no guarantees are provided about preserving keys after a compaction.
+
+- Measurement Block - This block stores the measurement name and a list of tag keys. Each tag key points to a tag value block.
+
+- Tag Value Block - This block stores a list of values for a tag key that points to a sorted list of series ids. The series ids are used instead of series keys because the IDs are much smaller than the keys and sorted integer sets can be operated on more quickly.
+
+- WAL Block - This block stores a list of series keys that have not been compacted yet.
+
+## Usage
+
+Initially, series are appended to a WAL block and the series information is held in-memory. Once this reaches a threshold, the WAL block is compacted to a set of series, measurement, & tag value blocks. The WAL is discarded and these new blocks will serve as the index and will be operated upon through an mmap.
+
+Series keys can be queried by measurement and tag value by first jumping to the appropriate measurement block. References to measurement blocks will be held in-memory once the files are read in. From the measurement block, a binary search on the tag keys will allow you to find a particular key. Once the key is found, you can jump to the key's tag value block and perform a binary search to find a particular value and a list of associated series IDs.
+
+_One future optimization may be to maintain an in-memory histogram to limit the number of pages searched during tag key and tag value traversal._
+
+### Tiered storage
+
+Once the series, measurement, and tag value blocks are written for a WAL block, a new WAL block can be started at the end of the file. Once this WAL block reaches a threshold size, it can be rewritten as a new set of series, measurement, and tag value blocks.
+
+These measurement blocks may duplicate measurement blocks created earlier by other WAL blocks so looking up measurements will involve first searching the new measurement block and then falling back to older measurement blocks and merging the data at query time.
+
+### Compaction
+
+As TSI files grow and measurements are duplicated, query time merging becomes more costly so a larger compaction will be necessary to consolidate all series, measurement, and tag value blocks into one.
+
+
+## Detailed Block Information
+
+### Series Blocks
+
+Series blocks contain a `uint32` block ID followed by a list of `uint32` local series IDs. Combining the block id and local series ID provides a globally addressable `uint64` series ID: `(blockID,seriesID)`
+
+The layout on disk for the block would be:
+
+```
+BLOCKID<uint32>
+KEYN<uint32>
```

Comment:
> The end of the block can be determined by looking at the last KEYPOS and jumping to the end of that key. The blocks will be scanned on initial load and block positions will be saved in-memory.

> The end of the block can be determined by looking at the last KEYPOS and jumping to the end of that key. The blocks will be scanned on initial load and block positions will be saved in-memory.

Got it. 

> New series sets will be appended (instead of doing a full compaction) so we'll need to have multiple series blocks.

Oh, I didn't realize there were two types of compactions.  I was assuming a compaction was a full rewrite.  

Are these statements accurate?
* Fast compactions convert WAL blocks to new segment blocks.  Duplicate measurement, tag, value blocks could be created by this process that would need to be deduped at search time?
* Full compactions rewrite the index to remove all duplicates